libraries/hotstuff/test/test_hotstuff_state.cpp

```diff
@@ -0,0 +1,160 @@
+#include <stdio.h>
+
+#include <boost/test/unit_test.hpp>
+#include <eosio/chain/types.hpp>
+
+#include <fc/io/cfile.hpp>
+
+#include <fc/exception/exception.hpp>
+#include <eosio/hotstuff/qc_chain.hpp>
+
+#include <fc/crypto/bls_signature.hpp>
+
+using std::cout;
+
+BOOST_AUTO_TEST_SUITE(test_hotstuff_state)
+
+const std::string file_path_1("temp_hs_safety");
+//const std::string file_path_2("temp_hs_liveness");
+
+BOOST_AUTO_TEST_CASE(write_safety_state_to_file) try {
+
+  eosio::hotstuff::hs_proposal_message hspm_1;
+  eosio::hotstuff::hs_proposal_message hspm_2;
+
+  hspm_1.block_id = eosio::chain::block_id_type("0b93846cf55a3ecbcd8f9bd86866b1aecc2e8bd981e40c92609ce3a68dbd0824"); //UX Network block #194217067
+  hspm_1.final_on_qc = eosio::chain::block_id_type();
+  hspm_1.phase_counter = 2;
+
+  eosio::hotstuff::view_number v_height = hspm_1.get_view_number();
+
+  hspm_2.block_id = eosio::chain::block_id_type("0b93846ba73bdfdc9b2383863b64f8f921c8a2379d6dde4e05bdd2e434e9392a"); //UX Network block #194217067
+  hspm_2.final_on_qc = eosio::chain::block_id_type();
+  hspm_2.phase_counter = 0;
+
+  fc::sha256 b_lock = eosio::hotstuff::get_digest_to_sign(hspm_2.block_id, hspm_2.phase_counter, hspm_2.final_on_qc);
+
+  eosio::hotstuff::safety_state ss;
+
+  ss.set_v_height(eosio::chain::name{""}, v_height);
+  ss.set_b_lock(eosio::chain::name{""}, b_lock);
+  
+  eosio::hotstuff::write_state(file_path_1, ss);
+
+  bool ok = true;
+
+  BOOST_CHECK_EQUAL(ok, true);
+
+} FC_LOG_AND_RETHROW();
+
+BOOST_AUTO_TEST_CASE(read_safety_state_from_file) try {
+
+  eosio::hotstuff::safety_state ss;
+
+  eosio::hotstuff::read_state(file_path_1, ss);
+
+  std::remove(file_path_1.c_str());
+
+  //test correct values
+  eosio::hotstuff::hs_proposal_message hspm_1;
+  eosio::hotstuff::hs_proposal_message hspm_2;
+
+  hspm_1.block_id = eosio::chain::block_id_type("0b93846cf55a3ecbcd8f9bd86866b1aecc2e8bd981e40c92609ce3a68dbd0824"); //UX Network block #194217067
+  hspm_1.final_on_qc = eosio::chain::block_id_type();
+  hspm_1.phase_counter = 2;
+
+  eosio::hotstuff::view_number v_height = hspm_1.get_view_number();
+
+  hspm_2.block_id = eosio::chain::block_id_type("0b93846ba73bdfdc9b2383863b64f8f921c8a2379d6dde4e05bdd2e434e9392a"); //UX Network block #194217067
+  hspm_2.final_on_qc = eosio::chain::block_id_type();
+  hspm_2.phase_counter = 0;
+
+  fc::sha256 b_lock = eosio::hotstuff::get_digest_to_sign(hspm_2.block_id, hspm_2.phase_counter, hspm_2.final_on_qc);
+
+  //std::pair<eosio::chain::view_number, fc::sha256> ss = get_safety_state(eosio::chain::name{""});
+
+  bool ok1 = ss.get_v_height(eosio::chain::name{""}) == v_height;
+  bool ok2 = ss.get_b_lock(eosio::chain::name{""}) == b_lock;
+
+  BOOST_CHECK_EQUAL(ok1, true);
+  BOOST_CHECK_EQUAL(ok2, true);
+
+} FC_LOG_AND_RETHROW();
+
+/*BOOST_AUTO_TEST_CASE(write_liveness_state_to_file) try {

```

Comment:
Why are these commented out?

Reply:
Maybe the liveness state file spec is not settled? In any case, we could open a new "IF: Persist hotstuff liveness state" issue, which would use the same state file read/write support code that's introduced by this PR. The commented-out liveness state file code in this PR can be left in to help the PR for that next issue.